Overhaul of dpctl documentation (#1619)
IntelPython/dpctl: docs/doc_sources/reference_guides/dpctl/tensor.data_types.rst

Make sure to select an appropriately capable device for an application that requires use of double
precision floating-point type.

Default complex floating-point data type
----------------------------------------

Like for the default real floating-point type, the default complex floating-point type depends on
capabilities of device. If the device support double precision real floating-point types, the default
complex floating-point type is :attr:`complex128`, otherwise :attr:`complex64`.

Querying default data types programmatically
--------------------------------------------

The data type can be discovered programmatically using Array API :ref:`inspection functions <dpctl_tensor_inspection>`:

.. code-block:: python

    from dpctl
    from dpctl import tensor

    device = dpctl.select_default_device()
    # get default data types for default-selected device
    default_types = tensor.__array_namespace_info__().default_dtypes(device)
    int_dt = default_types["integral"]
    ind_dt = default_types["indexing"]
    rfp_dt = default_types["real floating"]
    cfp_dt = default_types["complex floating"]

Type promotion rules
--------------------

> Type promotion rules govern behavior of array library when its function does not have
> dedicated implementation for the data type(s) of the input array(s).
> 
> In such a case, input arrays may be cast to data types for which dedicated implementation
> exists. This is what happens when function :data:`sin` is applied to array of integral values.

Reviewer: ```suggestion
Type promotion rules govern the behavior of an array library when a function does not have
a dedicated implementation for the data type(s) of the input array(s).

In such a case, input arrays may be cast to data types for which a dedicated implementation
exists. e.g., when :data:`sin` is applied to array of integral values.
```

Type promotion rules used in :py:mod:`dpctl.tensor` are consistent with the
Python Array API specification's `type promotion rules <https://data-apis.org/array-api/latest/API_specification/type_promotion.html>`_
for devices that support double precision floating-point type.

For devices that do not support double precision floating-point type, the type promotion rule is

> truncated by removing nodes corresponding to unsupported data types and edges that lead to them.

Reviewer: Is this entirely accurate? For instance, adding an array of type i8 to an array of type f4 would normally result in f8, but if f8 isn't supported, it results in f4. So the edge is less removed and more redirected to the nearest supported node as I see it.

Author: Perhaps this paragraph must be expanded. The graph itself is modified as described, but the graph from array API does not specify cross-kind promotion at all, and that needs to be spelled out here.

Reviewer: Yes, you're right, I only realized it afterward that it was specifically referring to the array API style of type promotion.
Maybe the best way to do it is to specify how we specifically go beyond the spec with regard to type promotion.
